Code/EnginePlugins/JoltPlugin/Components/Implementation/JoltBreakableSlabComponent.cpp

```diff
@@ -0,0 +1,1348 @@
+    return false;
+
+  return true;
+}
+
+ezUInt32 ezJoltBreakableSlabComponent::FindClosestShard(const ezVec3& vGlobalPosition) const
+{
+  const ezVec2 vLocalPos = (GetOwner()->GetGlobalTransform().GetInverse() * vGlobalPosition).GetAsVec2();
+  float fBestDistance = ezMath::HighValue<float>();
+
+  ezUInt32 uiShardIdx = ezInvalidIndex;
+
+  const ezUInt32 uiShards = m_Breakable.m_Shards.GetCount();
+  for (ezUInt32 i = 0; i < uiShards; ++i)
+  {
+    const auto& shard = m_Breakable.m_Shards[i];
+    if (shard.m_bShattered || shard.m_bDynamic)
+      continue;
+
+    const float fDistSqr = (shard.m_vCenterPosition - vLocalPos).GetLengthSquared();
+
+    if (fDistSqr > fBestDistance)
+      continue;
+
+    // if the shard is smaller than we are close to it, ignore it
+    if (fDistSqr > ezMath::Square(shard.m_fBoundingRadius))
+      continue;
+
+    fBestDistance = fDistSqr;
+    uiShardIdx = i;
+  }
+
+  return uiShardIdx;
+}
+
+void ezJoltBreakableSlabComponent::ShatterCellular(const ezVec3& vGlobalPosition, float fCellSize, const ezVec3& vImpulse, float fMakeDynamicRadius)
+{
+  if (!IsPointOnSlab(vGlobalPosition))
+    return;
+
+  const ezUInt32 uiShardIdx = FindClosestShard(vGlobalPosition);
+  if (uiShardIdx == ezInvalidIndex)
+    return;
+
+  if (m_ShatterPoints.IsEmpty())
+  {
+    ((ezJoltBreakableSlabComponentManager*)GetOwningManager())->m_RequireBreakage.Insert(GetHandle());
+  }
+
+  auto& pt = m_ShatterPoints.ExpandAndGetRef();
+  pt.m_vGlobalPosition = vGlobalPosition;
+  pt.m_vImpulse = vImpulse;
+  pt.m_fCellSize = fCellSize;
+  pt.m_uiShardIdx = uiShardIdx;
+  pt.m_fMakeDynamicRadius = fMakeDynamicRadius;
+  pt.m_uiAllowedBreakPatterns = (ezUInt8)ezBreakablePattern::Cellular;
+}
+
+void ezJoltBreakableSlabComponent::ShatterRadial(const ezVec3& vGlobalPosition, float fImpactRadius, const ezVec3& vImpulse, float fMakeDynamicRadius)
+{
+  if (!IsPointOnSlab(vGlobalPosition))
+    return;
+
+  const ezUInt32 uiShardIdx = FindClosestShard(vGlobalPosition);
+  if (uiShardIdx == ezInvalidIndex)
+    return;
+
+  if (m_ShatterPoints.IsEmpty())
+  {
+    ((ezJoltBreakableSlabComponentManager*)GetOwningManager())->m_RequireBreakage.Insert(GetHandle());
+  }
+
+  auto& pt = m_ShatterPoints.ExpandAndGetRef();
+  pt.m_vGlobalPosition = vGlobalPosition;
+  pt.m_vImpulse = vImpulse;
+  pt.m_fImpactRadius = fImpactRadius;
+  pt.m_fCellSize = 0.4f;
+  pt.m_uiShardIdx = uiShardIdx;
+  pt.m_fMakeDynamicRadius = fMakeDynamicRadius;
+  pt.m_uiAllowedBreakPatterns = (ezUInt8)ezBreakablePattern::Radial | (ezUInt8)ezBreakablePattern::Cellular;
+}
+
+void ezJoltBreakableSlabComponent::ShatterAll(float fShardSize, const ezVec3& vImpulse)
+{
+  m_fContactReportForceThreshold = 0.0f;
+
+  if (m_ShatterPoints.IsEmpty())
+  {
+    ((ezJoltBreakableSlabComponentManager*)GetOwningManager())->m_RequireBreakage.Insert(GetHandle());
+  }
+
+  auto& pt = m_ShatterPoints.ExpandAndGetRef();
+  pt.m_vGlobalPosition.Set(1024, 1024, 1024);
```

Comment:
Can we make this a constant like `ezInvalidIndex`? E.g. `ezShatterALl` ?

Reply:
yeah, but this is only internal, and only used in 2 places, so not that important atm